### ISSUE TITLE
Bug: [no-unnecessary-condition] false positive for nullish coalescing assignment of uninitialized variable

### ISSUE THREAD

**Issue by swandir** [typescript-eslint/typescript-eslint#6926]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=DYUwLgBAdgXNCuBbARiATgKChA-DgvBAAxA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
let n: number
n ??= 0
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Error should not be reported

### Actual Result

Error is reported

### Additional Info

When using `let name: type` variable declaration, TS track the variable initialization and only allows reading the variable after it's been assigned.

Conditional assignment to an uninitialized variable is allowed, even though it might be seen as reading and then assigning.

`no-unnecessary-condition` has a problem with that, as the variable can't be `null | undefined`, but only because TS won't allow reading it until it's assigned.

Similar to #6632, but for a different TS feature.

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/issues/6635

Please use the search before filing - it's the first issue that shows up when searching the rule name.

**Comment by swandir** [comment]
> #6635
> 
> Please use the search before filing - it's the first issue that shows up when searching the rule name.

Sorry if I'm mistaken, looks like a different issue though?

#6635 is about optional properties while this one is about uninitialized variables.

Will a fix for #6635 also fix the rule for the uninitialized variable case too?

**Comment by Josh-Cena** [comment]
Handling one does not handle the other (that one will be fixe by checking index signatures). I'm personally unsure how this one can be handled.

**Comment by Mitsunee** [comment]
This may or may not be an issue with typescript itself as `let n: number` should result in `n` being typed as `number | undefined` until confirmed to be initialized. I am currently updating my typescript ruleset and wanted to include this rule only to be met with this issue about 30 times all over my codebase.

**Comment by JoshuaKGoldberg** [comment]
Yeah we'll have to add logic to the rule to understand whether the value was maybe not initialized yet.

**Comment by Josh-Cena** [comment]
Duplicate of #4513
